File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: Optional[TreeNode]) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if not root:
            return None
       # head = root
        #Morris preorder traversal
        def rightmost_in_left_tree(node):
            node = node.left
            while(node.right):
                node = node.right
            return node
        head = TreeNode(-1)
        cur_list = head
        cur = root
        while(cur):
            cur_list.right = cur
            cur_list.left = None
            if cur.left != None:
                rightmost_in_left = rightmost_in_left_tree(cur)
                rightmost_in_left.right = cur.right
                cur = cur.left
            else:
                cur = cur.right
            cur_list = cur_list.right
        return root

File: 114-flatten-binary-tree-to-linked-list/114-flatten-binary-tree-to-linked-list.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void flatten(TreeNode root) {
        //@input a binary tree
        //@output linked list with the same treenode class, right is same to the next and it is a pre-order traversal.
        //morris pre-order traversal
        //@edge case 
        TreeNode dumb = new TreeNode(0);
        TreeNode cur = root;
        TreeNode pointer = dumb;
        while(cur != null){
            pointer.right = cur;
            pointer = cur;
            if(cur.left != null){
                TreeNode node = getRightMostNode(cur);
                node.right = cur.right;
                cur.right = null;
                TreeNode prev = cur;
                cur = cur.left;
                prev.left = null;
                
            }else{
                cur = cur.right;
            }
        }
        
        
        
        
        
    }
    public TreeNode getRightMostNode(TreeNode cur){
        TreeNode node = cur.left;
        while(node.right != null){
            node = node.right;
        }
        return node;
    }
    
}